File: backend/safety/health_check.py
import asyncio
import logging

from safety.rollback import rollback_manager

logger = logging.getLogger(__name__)


class HealthCheckLoop:
    """Background health check loop that monitors probes during experiments.

    Polls probes at a configurable interval and triggers automatic rollback
    when consecutive failures exceed the threshold.
    """

    def __init__(
        self,
        experiment_id: str,
        probes: list,
        interval: int = 10,
        failure_threshold: int = 3,
        on_failure: callable | None = None,
    ):
        self.experiment_id = experiment_id
        self.probes = probes
        self.interval = interval
        self.failure_threshold = failure_threshold
        self.on_failure = on_failure
        self._consecutive_failures = 0
        self._task: asyncio.Task | None = None
        self._stopped = asyncio.Event()
        self._results: list = []

    @property
    def results(self) -> list:
        return list(self._results)

    @property
    def is_running(self) -> bool:
        return self._task is not None and not self._task.done()

    def start(self) -> None:
        """Start the health check loop as a background task."""
        if self._task is not None:
            return
        self._stopped.clear()
        self._task = asyncio.create_task(self._run())
        logger.info(
            "Health check loop started for %s (interval=%ds, threshold=%d)",
            self.experiment_id,
            self.interval,
            self.failure_threshold,
        )

    async def stop(self) -> None:
        """Stop the health check loop."""
        if self._task is None:
            return
        self._stopped.set()
        try:
            await asyncio.wait_for(self._task, timeout=self.interval + 2)
        except (TimeoutError, asyncio.CancelledError):
            self._task.cancel()
        self._task = None
        logger.info("Health check loop stopped for %s", self.experiment_id)

    async def _run(self) -> None:
        """Main polling loop."""
        while not self._stopped.is_set():
            try:
                all_passed = await self._check_probes()
                if all_passed:
                    self._consecutive_failures = 0
                else:
                    self._consecutive_failures += 1
                    logger.warning(
                        "Health check failed for %s (%d/%d)",
                        self.experiment_id,
                        self._consecutive_failures,
                        self.failure_threshold,
                    )

                    if self._consecutive_failures >= self.failure_threshold:
                        logger.critical(
                            "Health check threshold reached for %s. Triggering rollback.",
                            self.experiment_id,
                        )
                        if self.on_failure:
                            await self.on_failure()
                        else:
                            await rollback_manager.rollback(self.experiment_id)
                        self._stopped.set()
                        return

                # Wait for interval or stop signal
                try:
                    await asyncio.wait_for(self._stopped.wait(), timeout=self.interval)
                    return  # Stopped
                except TimeoutError:
                    pass  # Continue polling

            except asyncio.CancelledError:
                return
            except Exception as e:
                logger.error("Health check loop error for %s: %s", self.experiment_id, e)
                self._consecutive_failures += 1

    async def _check_probes(self) -> bool:
        """Execute all probes and return True if all pass."""
        if not self.probes:
            return True

        all_passed = True
        for probe in self.probes:
            result = await probe.safe_execute()
            self._results.append(result)
            if not result.passed:
                all_passed = False
        return all_passed

File: backend-go/internal/safety/healthcheck.go
package safety

import (
	"context"
	"log"
	"sync"
	"time"
)

// HealthProbe is the interface that health check probes must implement
type HealthProbe interface {
	Execute(ctx context.Context) (passed bool, err error)
	Name() string
}

// HealthCheckLoop monitors probes during experiments and triggers rollback
// when consecutive failures exceed the threshold
type HealthCheckLoop struct {
	experimentID     string
	probes           []HealthProbe
	interval         time.Duration
	failureThreshold int
	onFailure        func()
	rollbackMgr      *RollbackManager

	mu                  sync.Mutex
	consecutiveFailures int
	running             bool
	cancel              context.CancelFunc
}

// NewHealthCheckLoop creates a new health check loop
func NewHealthCheckLoop(
	experimentID string,
	probes []HealthProbe,
	interval time.Duration,
	failureThreshold int,
	rollbackMgr *RollbackManager,
) *HealthCheckLoop {
	return &HealthCheckLoop{
		experimentID:     experimentID,
		probes:           probes,
		interval:         interval,
		failureThreshold: failureThreshold,
		rollbackMgr:      rollbackMgr,
	}
}

// Start begins the health check polling loop in a goroutine
func (hc *HealthCheckLoop) Start() {
	hc.mu.Lock()
	if hc.running {
		hc.mu.Unlock()
		return
	}
	hc.running = true

	ctx, cancel := context.WithCancel(context.Background())
	hc.cancel = cancel
	hc.mu.Unlock()

	log.Printf("Health check loop started for %s (interval=%v, threshold=%d)",
		hc.experimentID, hc.interval, hc.failureThreshold)

	go hc.run(ctx)
}

// Stop halts the health check loop
func (hc *HealthCheckLoop) Stop() {
	hc.mu.Lock()
	defer hc.mu.Unlock()

	if !hc.running {
		return
	}
	hc.running = false
	if hc.cancel != nil {
		hc.cancel()
	}
	log.Printf("Health check loop stopped for %s", hc.experimentID)
}

// IsRunning returns whether the loop is currently active
func (hc *HealthCheckLoop) IsRunning() bool {
	hc.mu.Lock()
	defer hc.mu.Unlock()
	return hc.running
}

func (hc *HealthCheckLoop) run(ctx context.Context) {
	ticker := time.NewTicker(hc.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			allPassed := hc.checkProbes(ctx)
			if allPassed {
				hc.consecutiveFailures = 0
				continue
			}

			hc.consecutiveFailures++
			log.Printf("Health check failed for %s (%d/%d)",
				hc.experimentID, hc.consecutiveFailures, hc.failureThreshold)

			if hc.consecutiveFailures >= hc.failureThreshold {
				log.Printf("Health check threshold reached for %s. Triggering rollback.",
					hc.experimentID)

				if hc.onFailure != nil {
					hc.onFailure()
				} else if hc.rollbackMgr != nil {
					hc.rollbackMgr.Rollback(hc.experimentID)
				}

				hc.Stop()
				return
			}
		}
	}
}

func (hc *HealthCheckLoop) checkProbes(ctx context.Context) bool {
	if len(hc.probes) == 0 {
		return true
	}

	for _, probe := range hc.probes {
		passed, err := probe.Execute(ctx)
		if err != nil || !passed {
			return false
		}
	}
	return true
}
